neuralop/training/trainer.py

```diff
@@ -130,173 +93,163 @@ def train(
-                    self.callbacks.on_before_loss(out=out)
-
-                loss = 0.0
-
-                if self.overrides_loss:
-                    loss += self.callbacks.compute_training_loss(
-                        out=out, **sample, amp_autocast=self.amp_autocast
-                    )
-                else:
-                    if self.amp_autocast:
-                        with amp.autocast(enabled=True):
-                            loss += training_loss(out, **sample)
-                    else:
-                        loss += training_loss(out, **sample)
-
-                if regularizer:
-                    loss += regularizer.loss
-
-                loss.backward()
-                del out
-
-                optimizer.step()
-                train_err += loss.item()
-
-                with torch.no_grad():
-                    avg_loss += loss.item()
-                    if regularizer:
-                        avg_lasso_loss += regularizer.loss
-
-                if self.callbacks:
-                    self.callbacks.on_batch_end()
-
-            if isinstance(scheduler, torch.optim.lr_scheduler.ReduceLROnPlateau):
-                scheduler.step(train_err)
-            else:
-                scheduler.step()
-
-            epoch_train_time = default_timer() - t1
-
-            train_err /= len(train_loader)
-            avg_loss /= n_samples
-            if regularizer:
-                avg_lasso_loss /= n_samples
-            else:
-                avg_lasso_loss = None
-
-            # collect info to log, message to print
-            if epoch % self.log_test_interval == 0:
-                if self.callbacks:
-                    self.callbacks.on_before_val(
-                        epoch=epoch,
-                        train_err=train_err,
-                        time=epoch_train_time,
-                        avg_loss=avg_loss,
-                        avg_lasso_loss=avg_lasso_loss,
-                    )
-                
-                all_errors = {}
+            if epoch % self.eval_interval == 0:
+                # evaluate and gather metrics across each loader in test_loaders
+                all_metrics = {}
                 for loader_name, loader in test_loaders.items():
-                    errors = self.evaluate(eval_losses, loader,
-                                           log_prefix=loader_name)                        
-                    all_errors.update(**errors)
+                    loader_metrics = self.evaluate(eval_losses, loader,
+                                            log_prefix=loader_name)                        
+                    all_metrics.update(**loader_metrics)
 
-                # print msg to console and optionally log to wandb
+                # log metrics at eval_interval
                 if self.verbose:
                     lr = None
-                    for pg in optimizer.param_groups:
+                    for pg in self.optimizer.param_groups:
```

Comment:
I'm wondering if we'd want to move this to the actual evaluation function (maybe with an additional is_last ?) so the user has the option to i) do it themselves and ii) log images too